pinax/api/viewsets.py

```diff
@@ -123,22 +123,46 @@ def parse_data(self):
             raise ErrorResponse(**self.error_response_kwargs(str(e), title="Invalid JSON", status=400))
 
     @contextlib.contextmanager
-    def validate(self, resource_class, obj=None):
+    def validate(self, resource_class, collection=False, obj=None):
+        """
+        Generator yields either a validated resource (collection=False)
+        or a resource generator callable (collection=True).
+
+        ValidationError exceptions resulting from subsequent (after yield)
+        resource manipulation cause an immediate ErrorResponse.
+        """
         data = self.parse_data()
         if "data" not in data:
             raise ErrorResponse(**self.error_response_kwargs('Missing "data" key in payload.', status=400))
-        if "attributes" not in data["data"]:
-            raise ErrorResponse(**self.error_response_kwargs('Missing "attributes" key in data.', status=400))
-        resource = resource_class()
+
+        if collection and not isinstance(data["data"], list):
+            raise ErrorResponse(**self.error_response_kwargs("Data must be in a list.", status=400))
+
         try:
-            resource.populate(data["data"], obj=obj)
-            yield resource
+            if collection:
+                def gen_func():
+                    for resource_data in data["data"]:
+                        yield self.validate_resource(resource_class, resource_data, obj)
+
+                yield gen_func
```

Comment:
I think we went the wrong way here. The generator idea was a little ill-conceived on my part. However, I think the fix is very easy and even cleaner. I am changing my mind after seeing the resulting code for the `validate` caller. Having to call the generator is a bit ugly IMO.

The new idea is to simply change this line to:

    yield (self.validate_resource(resource_class, resource_data, obj) for resource_data in data["data"])

The change to the caller is to remove the function call.